wip: refactor htlc switch

f321x opened the pull request:
See commit messages

wip, todo: 
* more unittests
* testing wallet db changes

f321x commented on electrum/lnworker.py:
```diff
@@ -2396,146 +2448,147 @@ def save_payment_info(self, info: PaymentInfo, *, write_to_disk: bool = True) ->
         payment_key = self._get_payment_key(payment_hash)
         status = self.received_mpp_htlcs.get(payment_key.hex())
-        return status and status.resolution == RecvMPPResolution.ACCEPTED
+        return status and status.resolution in (RecvMPPResolution.COMPLETE, RecvMPPResolution.SETTLING)
```
I think so, do you see anything specific case how this could go wrong?
Previously we had one state in which the mpp set was considered complete: `ACCEPTED`, now we have two states `COMPLETE` and `SETTLING`.

>  what if there is an ongoing incoming multi-channel MPP, and one channel gets force-closed?

The logic you referenced above should claim the pending htlc of the force closed channel, htlc_switch should settle the remaining htlcs of the open channels and remove them from the mpp set in `_fulfill_htlc_set()`. It will skip the force closed channel as its ctx cannot be updated anymore.
Once the closed channel is set redeemed `LNWallet.maybe_cleanup_mpp()` will remove the remaining htlcs and delete the mpp set so `htlc_switch` will stop iterating over it.

This logic should also be somewhat covered by the `extract_preimage` regtest:
https://github.com/spesmilo/electrum/blob/de7612b3b1b2262a76a92ae50f681b1ca3caa4c4/tests/regtest/regtest.sh#L406-L464

f321x replied:
In https://github.com/spesmilo/electrum/pull/10230/commits/ffb7c91f79448eecb37c5f37d4c593aa8ef08462 i changed `is_accepted_mpp()` to only return true if the set is in `SETTLING` state. This will make it respect `dont_settle_htlcs` and seems more conservative.

SomberNight replied:
moved this to https://github.com/spesmilo/electrum/pull/10318

ecdsa commented on electrum/wallet_db.py:
```diff
@@ -1157,6 +1158,23 @@ def _convert_version_60(self):
+            }
+            lightning_payments[rhash] = new_dataclass_type
+        self.data['seed_version'] = 61
+
```
If the wallet db upgrade is only motivated by the use of `dataclass`, then it might not be necessary.
Instead, the `stored_in` decorator can be attached to a constructor, see `UpdateHtlc`.

SomberNight replied:
The old db format did not store e.g. `creation_ts` and `expiry_delay`, so I think a db upgrade is needed.

SomberNight replied:
The old db format did not store e.g. `creation_ts` and `expiry_delay`, so I think a db upgrade is needed.

ecdsa replied:
oh right.

SomberNight commented on electrum/wallet_db.py:
```diff
@@ -1157,6 +1155,110 @@ def _convert_version_60(self):
+        # conversion of PaymentInformation
+        lightning_payments = self.data.get('lightning_payments', {})
+        for rhash, (amount_msat, direction, is_paid) in list(lightning_payments.items()):
+            new_dataclass_type = {
+                'rhash': rhash,
+                'amount_msat': amount_msat,
+                'direction': direction,
+                'status': is_paid,
+                'min_final_cltv_delta': 147,
+                'expiry_delay': 100 * 365 * 24 * 60 * 60,  # LN_EXPIRY_NEVER
+                'creation_ts': int(time.time()),
+            }
+            lightning_payments[rhash] = new_dataclass_type
```
It looks a bit wasteful to store the rhash twice in the DB. The inner dict rhash is implied by the outer key:
```
lightning_payments[rhash] = {
    'rhash': rhash,
    ...
}
```

f321x replied:
I removed the rhash attribute in the followup commit and store the PaymentInfo as tuple instead of dict.

f321x commented on electrum/lnworker.py:
```diff
@@ -106,34 +112,44 @@ class PaymentDirection(IntEnum):
+    rhash: str
     amount_msat: Optional[int]
+    # direction is being used with PaymentDirection and lnutil.Direction?
     direction: int
```
Yes that comment is incorrect, not sure how i got to this conclusion, maybe got confused by:
https://github.com/spesmilo/electrum/blob/de7612b3b1b2262a76a92ae50f681b1ca3caa4c4/electrum/lnworker.py#L1071-L1091

Removed in https://github.com/spesmilo/electrum/pull/10230/commits/6d17c8c91f33074a344e5e14bc4bad79ca22d87a

f321x commented on electrum/lnpeer.py:
```diff
@@ -2854,92 +2787,362 @@ async def htlc_switch_iteration():
+            * Fail whole set -> returns error code
+            * Settle whole set -> Returns preimage
+            * call callback (e.g. forwarding, hold invoice)
+        May modifies the mpp set in lnworker.received_mpp_htlcs (e.g. by setting its state to ACCEPTED).
```
Fixed in https://github.com/spesmilo/electrum/pull/10230/commits/6d17c8c91f33074a344e5e14bc4bad79ca22d87a

f321x commented on electrum/lnpeer.py:
```diff
@@ -2129,190 +2125,81 @@ def check_accepted_htlc(
-        if payment_hash.hex() in self.lnworker.dont_settle_htlcs:
-            return None, None
+        if len(htlc_set.htlcs) == 0:
+            self.lnworker.set_request_status(payment_hash, PR_PAID)
```
Yes right, in https://github.com/spesmilo/electrum/pull/10230/commits/ffb7c91f79448eecb37c5f37d4c593aa8ef08462 i changed it to set the request to `PR_PAID` already at the same time as setting the mpp set to `SETTLING` as at this point we can expect the mpp set to be claimed by us, either directly or onchain.

f321x commented on electrum/lnpeer.py:
```diff
@@ -2854,92 +2787,362 @@ async def htlc_switch_iteration():
+        elif mpp_set.resolution == RecvMPPResolution.EXPIRED:
+            return OnionFailureCode.MPP_TIMEOUT, None, None
+
+        if mpp_set.resolution == RecvMPPResolution.SETTLING:
+            # this is an ongoing forwarding, or a set that has not yet been fully settled (and removed).
+            # note the htlcs in SETTLING will not get failed automatically,
+            # even if timeout comes close, so either a forwarding failure or preimage has to be set
+            error_bytes, failure_message = self.lnworker.get_forwarding_failure(payment_key)
+            if failure_message or error_bytes:
+                # this was a forwarding set and it failed
+                self.lnworker.set_mpp_resolution(payment_key, RecvMPPResolution.FAILED)
+                return error_bytes or failure_message, None, None
```
yes the transition from `SETTLING` to `FAILED` feels 'wrong' however for the forwarding purpose i set the set to 'SETTLING' to prevent the set from getting failed in any other way than through a forwarding failure. We could also rename the state to `SETTLING_OR_FORWARDING` but that doesn't seem to elegant either.

f321x replied:
Added your comment in https://github.com/spesmilo/electrum/pull/10230/commits/1c59353d75e6e8cc4ffbc4976b088f3834b9c0de

f321x commented on electrum/lnutil.py:
```diff
@@ -505,7 +505,11 @@ class LNProtocolWarning(Exception):
 # set it a tiny bit higher for invoices as blocks could get mined
 # during forward path of payment
-MIN_FINAL_CLTV_DELTA_FOR_INVOICE = MIN_FINAL_CLTV_DELTA_ACCEPTED + 3
+MIN_FINAL_CLTV_DELTA_FOR_INVOICE = MIN_FINAL_CLTV_DELTA_ACCEPTED
```
this doesn't seem useful anymore, removed it in the followup commit

SomberNight commented on electrum/lnworker.py:
```diff
@@ -2425,138 +2458,137 @@ def save_payment_info(self, info: PaymentInfo, *, write_to_disk: bool = True) ->
+        if new_resolution in (RecvMPPResolution.COMPLETE, RecvMPPResolution.SETTLING):
+            # don't allow set to be completed if it is already failed,
+            # newly received htlcs should get failed when their set is already failed
+            assert mpp_status.resolution not in (RecvMPPResolution.EXPIRED, RecvMPPResolution.FAILED)
```
Looks good. Thanks!

f321x commented:
Rebased on master from HEAD https://github.com/spesmilo/electrum/pull/10230/commits/2fe9d1ae078a70463e3c91482fc90d8b32b10cd1 to new HEAD https://github.com/spesmilo/electrum/pull/10230/commits/64de57eaef54f7746394779cc727c315d516d23b

SomberNight commented on electrum/lnonion.py:
```diff
@@ -411,6 +445,33 @@ def process_onion_packet(
 
+def compare_trampoline_onions(trampoline_onions: List[Optional[ProcessedOnionPacket]]) -> bool:
+    """
+    compare values of trampoline onions except for the amount_to_forward as it is expected
```
`amt_to_forward` should also be equal among the inner/trampoline onions

f321x commented on electrum/lnworker.py:
```diff
@@ -3385,7 +3482,64 @@ def maybe_add_backup_from_tx(self, tx):
+                else:
+                    assert not trampoline_onion.are_we_final and outer_onion.are_we_final
+                    # trampoline forwarding
+                    await self._maybe_forward_trampoline(
```
TODO: `_maybe_forward_trampoline` is called on each htlc of the set, it should probably just be called once once the set is complete.

SomberNight replied:
note: this is a blocker -- should be fixed before merging

SomberNight replied:
Perhaps it is confusing that `_maybe_forward_trampoline` gets passed the `outer_onion` as a param, as it should only be called once for the whole set and all the outer_onions differ. From the outer_onion it only needs the `total_msat` field, which should match between all the outer_onions. It also separately gets the `trampoline_onion` - which should also match between the incoming htlcs.

f321x replied:
Fixed in the rebase and added a unittest `test_trampoline_mpp_consolidation_forwarding_amount` intended to catch something like this.

SomberNight commented on electrum/lnpeer.py:
```diff
@@ -3080,75 +2641,125 @@ async def htlc_switch(self):
+                assert callback is None and any_error is None
+                if self.lnworker.enable_htlc_settle:
+                    self._fulfill_htlc_set(payment_key, preimage)
+            if callback:
+                assert preimage is None and any_error is None
+                task = asyncio.create_task(callback())
```
Ah, looks like not even that is true. `_check_unfulfilled_htlc_set` should only return the cb for the first peer. Subsequent runs won't return it, due to `mpp_resolution` having been set to SETTLING.

f321x commented:
Rebased from https://github.com/spesmilo/electrum/pull/10230/commits/1226452c8b072614f107944172b2b99318a8c6cf to https://github.com/spesmilo/electrum/commit/9871b54e6d1c81daa137f8b432b4beb1e41c401a.

Restored behavior of mpp sets with hold invoice callback to pre PR way of not timing out htlcs exceeding `MIN_FINAL_CLTV_DELTA_ACCEPTED` if a callback is registered by setting the set to `SETTLING` once the callback got called and only failing it if the callback gets unregistered again with `LNWallet.unregister_hold_invoice`, otherwise the htlcs will get expired onchain (as before).
Added a unittest `test_hold_invoice_set_doesnt_get_expired` to specify this behavior more clearly.

Added some additional comparisons between the outer onion and the trampoline onion according to:
https://github.com/lightning/bolts/blob/9938ab3d6160a3ba91f3b0e132858ab14bfe4f81/04-onion-routing.md?plain=1#L547-L553

Reverted previous change of setting `amt_to_forward` equal to `total_msat` in the final trampoline onion. 
`compare_trampoline_onions` now only enforces equal `amt_to_forward` for forwarding sets while it accepts different `amt_to_forward` for sets we receive.

f321x commented on tests/test_lnpeer.py:
```diff
@@ -2120,6 +2448,74 @@ async def test_payment_trampoline_e2e_alice_t1_carol_t2_edward(self):
+            # test he should fail the incoming htlcs as their trampolines are not similar
+            first_payload = modified_hops_data[0].payload
+            amount_to_forward = first_payload['amt_to_forward']['amt_to_forward']
+            first_payload['amt_to_forward']['amt_to_forward'] = amount_to_forward - 1
```
Changed this in https://github.com/spesmilo/electrum/pull/10230/commits/ff19c200316f5c788d10b39ac8ac79cd45213b46

f321x commented:
Rebased from https://github.com/spesmilo/electrum/pull/10230/commits/49ccb06e4d075fc50a7c4b1cd1fc7632881cc716 to https://github.com/spesmilo/electrum/pull/10230/commits/610708fe763677bf046c807115b2be5c1bf76256.
Incoming trampoline payments are now handled in two stages, first the htlcs are grouped by the outer payment secret until sum(htlcs amounts) >= trampoline_onion.amt_to_forward, then the htlcs are collected in a single htlc set that has to reach total amount (as before). This allows to handle parts of multi trampoline payments independently, so parts of the payment can get retried without affecting the valid received parts.

f321x commented on electrum/trampoline.py:
```diff
@@ -310,6 +310,7 @@ def create_trampoline_onion(
             hops_data[i].payload["outgoing_node_id"] = {"outgoing_node_id": next_edge.node_id}
         # only for final
         if i == num_hops - 1:
+            payload["amt_to_forward"] = {"amt_to_forward": total_msat}
```
Updated the branch so it now handles the multi trampoline parts separately.

f321x commented on electrum/lnpeer.py:
```diff
@@ -2933,31 +3139,450 @@ async def wrapped_callback():
             onion_packet: OnionPacket, *,
             payment_hash: bytes,
-            onion_packet_bytes: bytes,
             is_trampoline: bool = False) -> ProcessedOnionPacket:
```
Ok, i am using sha256(onion_hash + payment_hash + is_trampoline) now in https://github.com/spesmilo/electrum/pull/10230/commits/2091b380d05a0b03259197073bbedd2d2cca09a4

SomberNight commented on electrum/lnonion.py:
```diff
@@ -157,6 +158,10 @@ def from_bytes(cls, b: bytes) -> 'OnionPacket':
 
+    @property
+    def onion_hash(self) -> bytes:
+        return sha256(self.to_bytes())
```
not sure how expensive this is, just noting that it could conceivably be `@cached_property`

f321x commented:
Rebased from https://github.com/spesmilo/electrum/pull/10230/commits/f49df4b6edb953b3ae44af602f1790f31d730185 to https://github.com/spesmilo/electrum/pull/10230/commits/b747046eca6b95e48293c41a3c16d569e9727e70.
Applies the suggested changes in the comments above.

f321x commented on electrum/lnpeer.py:
```diff
@@ -2933,31 +3141,453 @@ async def wrapped_callback():
+            * Fail whole set -> returns error code
+            * Settle whole set -> Returns preimage
+            * call callback (e.g. forwarding, hold invoice)
+        May modifies the mpp set in lnworker.received_mpp_htlcs (e.g. by setting its resolution to COMPLETE).
```
Fixed in https://github.com/spesmilo/electrum/pull/10230/commits/8feaaac34197e6a0de74a8836db741a8f1b3c057

f321x commented on electrum/lnworker.py:
```diff
@@ -3429,7 +3533,55 @@ def maybe_add_backup_from_tx(self, tx):
+                # trampoline forwarding
+                await self._maybe_forward_trampoline(
+                    payment_hash=any_mpp_htlc.htlc.payment_hash,
+                    inc_cltv_abs=any_mpp_htlc.htlc.cltv_abs,  # TODO: use max or enforce same value across mpp parts
```
Yes to me it also makes sense to use the min value as this is the closest time a htlc of the set will expire.
Added this in https://github.com/spesmilo/electrum/pull/10230/commits/8feaaac34197e6a0de74a8836db741a8f1b3c057

f321x commented on electrum/lnonion.py:
```diff
@@ -423,6 +458,49 @@ def process_onion_packet(
+        raise ValueError("nothing to compare")
+
+    if first_onion is None:
+        return all(onion is None for onion in trampoline_onions)
```
Added in https://github.com/spesmilo/electrum/pull/10230/commits/8feaaac34197e6a0de74a8836db741a8f1b3c057

f321x replied:
Added this additional check in https://github.com/spesmilo/electrum/pull/10230/commits/788e5c7b250e681b17de42f4ab7b210507bf86b9

f321x commented on electrum/lnworker.py:
```diff
@@ -2538,87 +2579,149 @@ def check_mpp_status(
+        # Payment key creation:
+        #   * for regular forwarded htlcs -> "scid.hex() + ':%d' % htlc_id" [htlc key]
+        #   * for trampoline forwarding -> "payment hash + payment secret from outer onion"
+        #   * for final htlcs (we are receiver) -> "payment hash + payment secret from (inner) onion"
```
Changed in https://github.com/spesmilo/electrum/pull/10230/commits/8feaaac34197e6a0de74a8836db741a8f1b3c057

f321x commented on electrum/lnworker.py:
```diff
@@ -3429,7 +3533,55 @@ def maybe_add_backup_from_tx(self, tx):
+        self,
+        payment_key: str, *,
+        processed_htlc_set: dict[ReceivedMPPHtlc, Tuple[ProcessedOnionPacket, Optional[ProcessedOnionPacket]]],
+    ) -> Optional[Callable[[], Awaitable[None]]]:
```
Fixed in https://github.com/spesmilo/electrum/pull/10230/commits/8feaaac34197e6a0de74a8836db741a8f1b3c057

f321x commented on electrum/lnpeer.py:
```diff
@@ -2933,31 +3141,453 @@ async def wrapped_callback():
+                    _log_fail_reason(f'incorrect trampoline onion {processed_onion=}\n{trampoline_onion=}')
+                    raise OnionRoutingFailure(code=OnionFailureCode.INVALID_ONION_PAYLOAD, data=b'\x00\x00\x00')
+
+            # note: the returned payment_key will contain the inner payment_secret
```
Fixed in https://github.com/spesmilo/electrum/pull/10230/commits/8feaaac34197e6a0de74a8836db741a8f1b3c057

f321x commented on electrum/lnworker.py:
```diff
@@ -2493,132 +2533,141 @@ def save_payment_info(self, info: PaymentInfo, *, write_to_disk: bool = True) ->
+
+    def is_complete_mpp(self, payment_hash: bytes) -> bool:
+        resolution = self.get_mpp_resolution(payment_hash)
+        return resolution and resolution in (RecvMPPResolution.COMPLETE, RecvMPPResolution.SETTLING)
```
Fixed this in https://github.com/spesmilo/electrum/pull/10230/commits/788e5c7b250e681b17de42f4ab7b210507bf86b9

f321x commented on electrum/lnpeer.py:
```diff
@@ -2126,20 +2125,100 @@ def check_accepted_htlc(
+                    amount_to_forward = processed_onion_packet.amt_to_forward
+                    error_data = amount_to_forward.to_bytes(8, byteorder="big")
+                e = OnionRoutingFailure(code=error_code, data=error_data or b'')
+                error_bytes = e.to_wire_msg(onion_packet, self.privkey, local_height)
```
Changed it to use the trampoline onions amt_to_forward in https://github.com/spesmilo/electrum/pull/10230/commits/788e5c7b250e681b17de42f4ab7b210507bf86b9

SomberNight commented on tests/test_lnpeer.py:
```diff
@@ -2121,6 +2672,106 @@ async def test_payment_trampoline_e2e_alice_t1_carol_t2_edward(self):
+                session_key,
+                hops_data,
+                **kwargs
+            )
```
the comment is a bit confusing: it returns an onion with modified `hops_data`
EDIT: well, for one HTLC.  but not for the other one.

SomberNight commented:
13e856b01b397bfd1564640f21470d18c1e56336 looks good
Only outstanding comment is basically when to automatically fail hold-invoice htlcs in relation to expiry.

SomberNight commented on tests/test_lnpeer.py:
```diff
@@ -1970,6 +1970,71 @@ async def f():
+                }
+
+            preimage = os.urandom(32)
+            lnaddr, pay_req = self.prepare_invoice(w2, payment_preimage=preimage, min_final_cltv_delta=244)
```
why 244? why not the default?
is it because 20 lines below you mine a few blocks, and the HTLC would get failed due to expiry getting close?

f321x commented on electrum/commands.py:
```diff
@@ -1462,15 +1465,15 @@ async def cancel_hold_invoice(self, payment_hash: str, wallet: Abstract_Wallet =
-        del wallet.lnworker.dont_settle_htlcs[payment_hash]
         result = {
             "cancelled": payment_hash
         }
```
Previously when `dont_settle_htlcs` was used for the hold invoice cli I wanted to prevent settling an htlc even if the preimage was already stored. 
With the loop i intended to prevent the htlc from accidentally getting settled if the preimage was already available by only removing the payment from `dont_settle_htlcs` after the htlcs got failed.
In practice however this could not really happen with the hold invoice cli (unless you somehow get the preimage in lnworker, e.g. through a plugin), now the hold invoice cli just doesn't try to prevent this anymore at all and will immediately settle if a preimage becomes available, so this 'race' isn't a concern anymore.
However i still let the while loop in there so the command blocks until the htlcs really got failed, i think its more honest to show the user 'cancelled' only once the htlcs really got cancelled. 
But yeah the comment is a bit misleading.

f321x commented on electrum/lnpeer.py:
```diff
@@ -2752,84 +2674,113 @@ async def htlc_switch(self):
         for chan_id, chan in self.channels.items():
+            # In this loop, an item (value) of chan.unfulfilled_htlcs may goes through 2 stages:
+            # - 1. not handled yet (new htlc): onion_packet_hex
+            # - 2. waiting for removal: None, will be cleaned up once irrevocably removed.
```
Yes i don't see why we would need to wait either, removed this in https://github.com/spesmilo/electrum/pull/10230/commits/74db8708b53501c6f3044e07dbbb5270607c77e5

SomberNight commented on electrum/lnpeer.py:
```diff
@@ -2752,84 +2674,105 @@ async def htlc_switch(self):
+                        error_bytes=error_bytes,
+                    )
+                finally:
+                    del unfulfilled[htlc_id]
```
How about something like this?

```diff
diff --git a/electrum/lnpeer.py b/electrum/lnpeer.py
index 49958ea33..3d029b1f5 100644
--- a/electrum/lnpeer.py
+++ b/electrum/lnpeer.py
@@ -2706,9 +2706,17 @@ class Peer(Logger, EventListener):
                     continue
 
                 htlc = chan.hm.get_htlc_by_id(REMOTE, htlc_id)
-                onion_packet = None
                 try:
                     onion_packet = self._parse_onion_packet(onion_packet_hex)
+                except OnionParsingError as e:
+                    self.fail_malformed_htlc(
+                        chan=chan,
+                        htlc_id=htlc.htlc_id,
+                        reason=e,
+                    )
+                    del unfulfilled[htlc_id]
+                    continue
+                try:
                     processed_onion_packet = self._process_incoming_onion_packet(
                         onion_packet,
                         payment_hash=htlc.payment_hash,
@@ -2725,22 +2733,29 @@ class Peer(Logger, EventListener):
                         htlc=htlc,
                         unprocessed_onion_packet=onion_packet_hex,  # outer onion if trampoline
                     )
-                except OnionParsingError as e:
+                except OnionParsingError as e:  # could be raised when parsing inner trampoline onion
                     self.fail_malformed_htlc(
                         chan=chan,
                         htlc_id=htlc.htlc_id,
                         reason=e,
                     )
-                except OnionRoutingFailure as e:
-                    assert onion_packet, "raise OnionParsingError if onion parsing failed"
+                except Exception as e:
                     # Fail the htlc directly if it fails to pass these tests, it will not get added to a htlc set.
                     # https://github.com/lightning/bolts/blob/14272b1bd9361750cfdb3e5d35740889a6b510b5/04-onion-routing.md?plain=1#L388
-                    error_bytes = e.to_wire_msg(onion_packet, self.privkey, self.network.get_local_height())
+                    reraise = False
+                    if isinstance(e, OnionRoutingFailure):
+                        orf = e
+                    else:
+                        orf = OnionRoutingFailure(code=OnionFailureCode.TEMPORARY_NODE_FAILURE, data=b'')
+                        reraise = True  # propagate this out, as this might suggest a bug
+                    error_bytes = orf.to_wire_msg(onion_packet, self.privkey, self.network.get_local_height())
                     self.fail_htlc(
                         chan=chan,
                         htlc_id=htlc.htlc_id,
                         error_bytes=error_bytes,
                     )
+                    if reraise:
+                        raise
                 finally:
                     del unfulfilled[htlc_id]

```

SomberNight replied:
Though note that even `fail_htlc`/`fail_malformed_htlc` could raise

f321x commented on electrum/lnpeer.py:
```diff
@@ -2752,84 +2674,119 @@ async def htlc_switch(self):
+        #    The sets are shared between all peers, so each peers htlc_switch acts on the same sets.
+        for payment_key, htlc_set in list(self.lnworker.received_mpp_htlcs.items()):
+            any_error, preimage, callback = self._check_unfulfilled_htlc_set(payment_key, htlc_set)
+            assert bool(any_error) + bool(preimage) + bool(callback) <= 1, f"{any_error=}{preimage=}{callback=}"
```
Improved in https://github.com/spesmilo/electrum/pull/10230/commits/4d397b95ca0ec3669190cb2e0997065076d491fc

f321x commented:
Rebased from https://github.com/spesmilo/electrum/commit/a6b0a03548b5efae47ed0c40e5b7e06f5cf33ddd to https://github.com/spesmilo/electrum/commit/472b77c783e3a510fe483909e54421cb65b2910e.
Just squashes the first two commits
